google-cloud-storage/src/test/java/com/google/cloud/storage/StorageImplMockitoTest.java

```diff
@@ -0,0 +1,583 @@
+          .put('/', "%2F")
+          .put(':', "%3A")
+          .put(';', "%3B")
+          .put('=', "%3D")
+          .put('?', "%3F")
+          .put('@', "%40")
+          .put('[', "%5B")
+          .put(']', "%5D")
+          // In addition to [a-zA-Z0-9], these chars should not be URI-encoded:
+          .put('-', "-")
+          .put('_', "_")
+          .put('.', ".")
+          .put('~', "~")
+          .build();
+
+  private static final String ACCOUNT = "account";
+  private static PrivateKey privateKey;
+  private static PublicKey publicKey;
+
+  private StorageOptions options;
+  private StorageRpcFactory rpcFactoryMock;
+  private StorageRpc storageRpcMock;
+  private Storage storage;
+
+  private Blob expectedBlob1, expectedBlob2, expectedBlob3;
+  private Bucket expectedBucket1, expectedBucket2, expectedBucket3;
+
+  @BeforeClass
+  public static void beforeClass() throws NoSuchAlgorithmException, InvalidKeySpecException {
+    KeyFactory keyFactory = KeyFactory.getInstance("RSA");
+    EncodedKeySpec privateKeySpec =
+        new PKCS8EncodedKeySpec(BaseEncoding.base64().decode(PRIVATE_KEY_STRING));
+    privateKey = keyFactory.generatePrivate(privateKeySpec);
+    EncodedKeySpec publicKeySpec =
+        new X509EncodedKeySpec(BaseEncoding.base64().decode(PUBLIC_KEY_STRING));
+    publicKey = keyFactory.generatePublic(publicKeySpec);
+  }
+
+  @Before
+  public void setUp() {
+    rpcFactoryMock = mock(StorageRpcFactory.class);
+    storageRpcMock = mock(StorageRpc.class);
+    when(rpcFactoryMock.create(any(StorageOptions.class))).thenReturn(storageRpcMock);
+    options =
+        StorageOptions.newBuilder()
+            .setProjectId("projectId")
+            .setClock(TIME_SOURCE)
+            .setServiceRpcFactory(rpcFactoryMock)
+            .setRetrySettings(ServiceOptions.getNoRetrySettings())
+            .build();
+  }
+
+  private void initializeService() {
+    storage = options.getService();
+    initializeServiceDependentObjects();
+  }
+
+  private void initializeServiceDependentObjects() {
+    expectedBlob1 = new Blob(storage, new BlobInfo.BuilderImpl(BLOB_INFO1));
+    expectedBlob2 = new Blob(storage, new BlobInfo.BuilderImpl(BLOB_INFO2));
+    expectedBlob3 = new Blob(storage, new BlobInfo.BuilderImpl(BLOB_INFO3));
+    expectedBucket1 = new Bucket(storage, new BucketInfo.BuilderImpl(BUCKET_INFO1));
+    expectedBucket2 = new Bucket(storage, new BucketInfo.BuilderImpl(BUCKET_INFO2));
+    expectedBucket3 = new Bucket(storage, new BucketInfo.BuilderImpl(BUCKET_INFO3));
+  }
+
+  @Test
+  public void testUploadNonExistentFile() {
+    initializeService();
+    String fileName = "non_existing_file.txt";
+    try {
+      storage.upload(BLOB_INFO1, Paths.get(fileName));
+      Assert.fail();
+    } catch (IOException e) {
+      assertEquals(NoSuchFileException.class, e.getClass());
+      assertEquals(fileName, e.getMessage());
+    }
+  }
+
+  @Test
+  public void testUploadDirectory() throws IOException {
```

Comment:
If you are converting existing tests into mockito then it would be better to use `dothrow().when().method()` to throw predefined exception something like below
```
  @Test
  public void testUploadDirectory() throws IOException {
    initializeService();
    Path dir = Files.createTempDirectory("unit_");
    String exceptionMessage = dir + " is a directory";
    doThrow(new StorageException(exceptionMessage))
        .when(storageRpcMock)
        .upload(...);
    try {
      storage.upload(BLOB_INFO1, dir);
    } catch (StorageException expected) {
      assertTrue(expected.getMessage().contains(exceptionMessage));
    }
  }
```

Reply:
testUploadWithException() does as you suggested.
testUploadDirectory() verifies exception thrown form upload() method. No mock is needed here.